docs: Improve liquidity module spec for concepts, state, README (#378)
tendermint/liquidity: x/liquidity/spec/01_concepts.md

## Liquidity Module

The liquidity module is a module that can be used on any Cosmos SDK based applications. It implements decentralized exchange (DEX) that serves liquidity providing and coin swap functions. Anyone can create a liquidity pool with a pair of coins, provide liquidity by depositing reserve coins into the liquidity pool, and trade coins using the liquidity pool. All the logics are designed in a way that always protect the pool investors.

## Liquidity Pool

A liquidity pool is a coin reserve that contains two different types of coins in a trading pair. It has to be unique. A liquidity provider can be anyone (permissionless) who provides liquidity by depositing reserve coins into the pool. He/she earns the accumulated swap fees with respect to their pool share. Pool share is represented as possession of pool coins. All matchable swap requests are expected to be executed and unmatched ones are removed.
## Equivalent Swap Price Model (ESPM)

> The liquidity module is a Cosmos SDK implementation of an AMM system with a novel economic model called the Equivalent Swap Price Model (ESPM). The key distinguishing feature of the ESPM model from the Constant Product Market Maker (CPMM) model (e.g. Uniswap) is the implementation of a hybrid system. This system combines an orderbook model exchange and a simple liquidity pool model by which the order book is governed by a set of order rules and execution is performed in batches. In the ESPM, the pool price is always equal to the last swap price, thus reducing opportunities for arbitrage.

Author: Thanks for the grammar tips. That is really helpful!

> The ESPM model is intended to provide protection against price volatility, transaction ordering vulnerabilities, and losses due to arbitrage, which are widely known vulnerabilities in AMMs such as Uniswap.

Author: I meant to specify all of those three. Those vulnerabilities (protection against price volatility, transaction ordering , losses due to arbitrage) are widely known in AMM Ethereum based projects.

## Batch Execution

> The liquidity module uses batch execution methodology. Making deposit, withdrawal, or swap orders are accumulated in a liquidity pool for a pre-defined period, which can be one or more blocks in length. Orders are then added to the pool and executed at the end of the batch. The size of each batch is changeable by using governance parameter called `UnitBatchSize`.

Author: Thanks for saying that :) I will make that modification.

## Price Discovery

Swap prices in liquidity pools are determined by the current pool coin reserves and the requested swap amount. Arbitrageurs buy or sell coins in liquidity pools to gain instant profit that results in real-time price discovery of liquidity pools.

## Escrow Process

The liquidity module uses module account that acts as an escrow account that holds and releases the coin amount during batch execution.

## Fees
### PoolCreationFee

> The liquidity module has `PoolCreationFee` parameter that is paid upon pool creation. This param is defined at genesis and it can be updated via the governance proposal. The purpose of this fee is to prevent from creating useless pools and making limitless transactions. This fee goes to community fund.

Author: > instead of Latin via, let's instead use "by using" (or "using)

That is good point. I tend to use word "via" and I didn't know it is Latin. 

> who is the actor who is being prevented? I used "users" but if we have a better term, we can use

I guess we can use "users". It can be retail traders or liquidity participants maybe?

## Liquidity Module

The liquidity module is a module that can be used on any Cosmos SDK based applications. It implements decentralized exchange (DEX) that serves liquidity providing and coin swap functions. Anyone can create a liquidity pool with a pair of coins, provide liquidity by depositing reserve coins into the liquidity pool, and trade coins using the liquidity pool. All the logics are designed in a way that always protect the pool investors.

## Liquidity Pool

A liquidity pool is a coin reserve that contains two different types of coins in a trading pair. It has to be unique. A liquidity provider can be anyone (permissionless) who provides liquidity by depositing reserve coins into the pool. He/she earns the accumulated swap fees with respect to their pool share. Pool share is represented as possession of pool coins. All matchable swap requests are expected to be executed and unmatched ones are removed.
## Equivalent Swap Price Model (ESPM)

> The liquidity module is a Cosmos SDK implementation of an AMM system with a novel economic model called the Equivalent Swap Price Model (ESPM). The key distinguishing feature of the ESPM model from the Constant Product Market Maker (CPMM) model (for example, Uniswap) is the implementation of a hybrid system. This system combines an orderbook model exchange with a simple liquidity pool model that governs the order book with a set of order rules and performs execution in batches. In the ESPM, the pool price is always equal to the last swap price which reduces opportunities for arbitrage.

Author: ESPM AMM model combines batch execution, order books and limit orders with an AMM to significantly outperform the CPMM model as seen on platforms like Uniswap.

> The liquidity module is a Cosmos SDK implementation of an AMM system with a novel economic model called the Equivalent Swap Price Model (ESPM). The key distinguishing feature of the ESPM model from the Constant Product Market Maker (CPMM) model (for example, Uniswap) is the implementation of a hybrid system. This system combines an orderbook model exchange with a simple liquidity pool model that governs the order book with a set of order rules and performs execution in batches. In the ESPM, the pool price is always equal to the last swap price which reduces opportunities for arbitrage.

Author: I think we should leave the sentence `The liquidity module is a Cosmos SDK implementation of an AMM system with a novel economic model called the Equivalent Swap Price Model (ESPM).` as is. CPMM is one of the AMM types (models) and it is different model from ESPM. 

I don't know if you already read our LitePaper. But here is [the link](https://github.com/tendermint/liquidity/blob/develop/doc/LiquidityModuleLightPaper_EN.pdf) to have more context about ESPM model.

> The liquidity module is a Cosmos SDK implementation of an AMM system with a novel economic model called the Equivalent Swap Price Model (ESPM). The key distinguishing feature of the ESPM model from the Constant Product Market Maker (CPMM) model (for example, Uniswap) is the implementation of a hybrid system. This system combines an orderbook model exchange with a simple liquidity pool model that governs the order book with a set of order rules and performs execution in batches. In the ESPM, the pool price is always equal to the last swap price which reduces opportunities for arbitrage.

Reviewer: good description, I think the content is clear now

The ESPM model is intended to provide protection against price volatility, transaction ordering vulnerabilities, and losses due to arbitrage. AMMs such as Uniswap do not provide protection against these whatever they are.

## Batch Execution

The liquidity module uses a batch execution methodology. Deposits, withdrawals, and swap orders are accumulated in a liquidity pool for a pre-defined period that is one or more blocks in length. Orders are then added to the pool and executed at the end of the batch. The size of each batch is configured by using the `UnitBatchSize` governance parameter.

## Price Discovery

Swap prices in liquidity pools are determined by the current pool coin reserves and the requested swap amount. Arbitrageurs buy or sell coins in liquidity pools to gain instant profit that results in real-time price discovery of liquidity pools.

## Escrow Process

The liquidity module uses a module account that acts as an escrow account. The module account holds and releases the coin amount during batch execution.

## Refund 

> The liquidity module has refund functions when deposit, withdraw, or swap batch states are not successfully executed.

Reviewer: ```suggestion
The liquidity module has refund functions when deposit, withdraw, or swap batch transactions are not successfully executed.
```

Reviewer: I think we want "transactions" here... not states?

Author: Technically states are right term in this case. Users send transactions and the transactions are accumulated in a batch state.

> Read [the logic](https://github.com/tendermint/liquidity/blob/e8ab2f4d75079157d008eba9f310b199573eed28/x/liquidity/keeper/batch.go#L83-L127) to get more context.

Reviewer: ```suggestion
Read [the batch transaction logic](https://github.com/tendermint/liquidity/blob/e8ab2f4d75079157d008eba9f310b199573eed28/x/liquidity/keeper/batch.go#L83-L127) in the code for more context.
```

> ## Fees

Reviewer: ```suggestion
## Fees

You set liquidity module fees for pool creation, withdrawal, and swap. 

```

Reviewer: best practice is to provide context for each section so we don't end up with stacked headings

### PoolCreationFee

> The liquidity module `PoolCreationFee` parameter is paid upon pool creation. This param is defined at genesis and can be updated by using the governance proposal. The purpose of this fee is to prevent users from creating useless pools and making limitless transactions. The funds from this fee go to the community fund.

Reviewer: ```suggestion
The liquidity module pool creation fee set by the `PoolCreationFee` parameter is paid on pool creation. This parameter is defined at genesis and can be updated by using the governance proposal. The purpose of this fee is to prevent users from creating useless pools and making limitless transactions. The funds from this fee go to the community fund.
```
